### PR TITLE
Add a gizmo-based overlay to show UI node outlines (Adopted)

### PR DESCRIPTION
**Pull request by pablo-lua**
# Objective

- This is an adopted version of #10420
- The objective is to help debugging the Ui layout tree with helpful outlines, that can be easily enabled/disabled

## Solution

- Like #10420, the solution is using the bevy_gizmos in outlining the nodes

---

### Added
- Added debug_overlay mod to `bevy_dev_tools`
- Added bevy_ui_debug feature to `bevy_dev_tools`

## How to use
- The user must use `bevy_dev_tools` feature in TOML
- The user must use the plugin UiDebugPlugin, that can be found on `bevy::dev_tools::debug_overlay`
- Finally, to enable the function, the user must set `UiDebugOptions::enabled` to true
Someone can easily toggle the function with something like:

```rust
fn toggle_overlay(input: Res<ButtonInput<KeyCode>>, options: ResMut<UiDebugOptions>) {
   if input.just_pressed(KeyCode::Space) {
      // The toggle method will enable if disabled and disable if enabled
      options.toggle();
   }
}
```

Note that this feature can be disabled from dev_tools, as its in fact behind a default feature there, being the feature bevy_ui_debug.

# Limitations
Currently, due to limitations with gizmos itself, it's not possible to support this feature to more the one window, so this tool is limited to the primary window only.

# Showcase

image
Ui example with debug_overlay enabled

image
And disabled

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
> Do we keep the keybinding config?

IMO no, this doesn't justify the complexity here (for now). Just make sure there's a good way for users to toggle it from their own code.

> Is it a good idea to move this to bevy_gizmos instead of bevy_ui?

Definitely belongs in bevy_ui for now. If the [plans regarding a bevy_debug_tools](https://discord.com/channels/691052431525675048/1192567920346271755/1192629117112307752) first-party crate come to fruition, this is another prime candidate for inclusion and should be moved there.

**Comment by pablo-lua** [comment]
> IMO no, this doesn't justify the complexity here (for now)

Agreed, for now I'll remove the key binding, the user can easily enough enable or disable by setting DebugOptions::enabled = true / false.

**Comment by 66OJ66** [comment]
It might be worth adding a toggle so that only `Nodes` which are children of a Visible root node have outlines applied.
(not too sure what the best approach would be though)

Reason for this: some Apps will create all their UI elements upfront under separate root nodes, then just show/hide these elements by changing the root node Visibility settings.

In this scenario, users could potentially see 100s of overlaid outlines which would make this new feature harder to use

**Comment by pablo-lua** [comment]
Added this, for now, It will hide the outlines of anything that is not visible judging by ViewVisibility component, let me know if you guys think that there is another better approach in this case

**Comment by pablo-lua** [comment]
Remembered about UiScale, I'll have a little look at it another hour, but this is a error so I'll throw this in draft until I fix it

**Comment by pablo-lua** [comment]
With this fixed, only thing that matters for now is adding some docs, I'm not very good at this, will try tomorrow 👍

File: crates/bevy_ui/src/debug_overlay/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/debug_overlay/mod.rs:220]:
```suggestion
/// Disabling the plugin will give you back gizmo control.
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Can you extend our main UI example demonstrating how to use this? Should be pretty simple as a change, and it's good for both discovery and easy validation.

**Comment by pablo-lua** [comment]
Added to the Ui examples, one must run `cargo run --example --features bevy_ui_debug`  to see the feature. One note is that its impossible to run this without fixing #11320, so its required to fix this at least temporarily to run this locally

**Comment by JMS55** [comment]
I think this should go in bevy_dev_tools https://github.com/bevyengine/bevy/pull/11341

File: examples/ui/ui.rs
Comment (mockersf) [inline examples/ui/ui.rs:99]: you could display a text if the feature is not enabled to say "try enabling feature `bevy_ui_debug`"

**Comment by pablo-lua** [comment]
> I think this should go in bevy_dev_tools 


Agreed, but should we Block this until the bevy tools get merged (#11341)?
(I think we should, but that depends on the PR situation)

**Comment by pablo-lua** [comment]
After the big changes on Ui, I noticed that some errors showed up, fixing them now 🔧

**Comment by matiqo15** [comment]
> should we Block this until the bevy tools get merged (#11341)?

I don't think there is a need for that. We can move it later - it's probably going to take some time to get #11341 merged.

**Comment by pablo-lua** [comment]
After some digging found out that this will going to be a hard migration, So I want to collect you guys opinion
The major problem is, we spawn a Debug Camera, which the user is not aware of. This wouldn't be a problem before, because we disabled the ui on that camera, but now every camera is a potential camera for ui (because we removed UiCameraOptions), so this would lead the user to add the TargetCamera to all nodes (at least, all the root nodes), even if they only spawned one camera and I think the easiest way to fix this behavior is by adding again the UiCameraOptions to skip this debug camera in specific (Correct me if I'm wrong on this aspect)

Now, something that must be decided: What we do about the different windows Ui? I think the easiest way to manage this type of thing is by spawning an different debug camera to each window (or by spawning a different camera at least for each target camera that has a different window itself), but what are you guys opinion on this design?

**Comment by pablo-lua** [comment]
Created #11377 to discuss this problem

**Comment by pablo-lua** [comment]
Merging #10342 would be the ideal solution to this problem, enabling and disabling only the debug line in specific.

**Comment by alice-i-cecile** [comment]
Linked PR is merging: no longer blocked :)

**Comment by pablo-lua** [comment]
Found a new panic, for now #11420 is blocking.

**Comment by pablo-lua** [comment]
Hmmm, dont quite get the Dependencies error, didn't changed anything in dependencies...

**Comment by matiqo15** [comment]
> Hmmm, dont quite get the Dependencies error, didn't changed anything in dependencies...

Seems to be failing since 35ac1b1

**Comment by pablo-lua** [comment]
Got it, thanks. Well, I'll open this PR for review, everything is working on new updates.

**Comment by pablo-lua** [comment]
Forgot to say: Due to the limitations with windows of bevy_gizmos, I limited this tool to the primary window only, I don't have any idea in particular of how to do this work on the currently API 🤔

**Comment by matiqo15** [comment]
> Forgot to say: Due to the limitations with windows of bevy_gizmos, I limited this tool to the primary window only, I don't have any idea in particular of how to do this work on the currently API 🤔

We should mention it in docs

**Comment by pablo-lua** [comment]
I'll throw it in a draft until I move this to the new `bevy_dev_tools` (see #11341)

File: crates/bevy_dev_tools/Cargo.toml
Comment (pablo-lua) [inline crates/bevy_dev_tools/Cargo.toml:13]: I'm wondering if I should do all the libs added here optional by feature flag, they might be used in the future by default, but at least now they are optional

File: crates/bevy_internal/Cargo.toml
Comment (pablo-lua) [inline crates/bevy_internal/Cargo.toml:156]: I did so when someone add this feature to the TOML, it will auto add bevy_dev_tools feature too. I'm not sure if that's a good idea or not.

**Review by matiqo15 — APPROVED** [review]

File: Cargo.toml
Comment (alice-i-cecile) [inline Cargo.toml:153]: IMO we shouldn't expose these features at the `bevy` level at all: instead, we can let users control them at the debug-tools crate level.

File: crates/bevy_dev_tools/src/debug_overlay/inset.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/debug_overlay/inset.rs:10]: Doc comments about why this exists (and why it exists here) would be useful.

File: Cargo.toml
Comment (pablo-lua) [reply Cargo.toml:153]: So, I'll remove it and for now, and we wait for #12368 to be capable of disable this?

File: crates/bevy_dev_tools/src/debug_overlay/inset.rs
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/debug_overlay/inset.rs:16]: Extracting this into an EPSILON associated constant would be good.

Or TBH we could just refactor this as a free `approx_eq(f32, f32) -> bool` function and I think this would be simpler and reduce the temptation for reuse.

File: crates/bevy_internal/Cargo.toml
Comment (alice-i-cecile) [reply crates/bevy_internal/Cargo.toml:156]: See comment above: we should just cut the top-level feature IMO.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
We should cut the top-level feature, and just feature flag this in an on-by-default feature in bevy_dev_tools. You may need to add bevy_dev_tools to Bevy's dev-deps to get the example to work by default.

Other concerns are just nits.

File: Cargo.toml
Comment (matiqo15) [reply Cargo.toml:153]: I would also not add a plugin by default, and let user specifically add it themself.

File: crates/bevy_dev_tools/src/debug_overlay/inset.rs
Comment (pablo-lua) [reply crates/bevy_dev_tools/src/debug_overlay/inset.rs:16]: I think I'll just remove this trait if I can, a simple function is probably better here.

**Comment by pablo-lua** [comment]
I removed the feature from top level features. Added `bevy_ui_debug` as a feature of bevy_dev_tools, and now you have to add the plugin yourself, like in the example, but I'm not sure if thats alright, so I could use of another opinion on what we have now. I'll fix the CI now btw.

**Comment by alice-i-cecile** [comment]
Hmm, if we add a dev-dependency on `bevy` in the `bevy` `Cargo.toml` with `bevy_debug_tools`  enabled does that work?

This may still be better, as it prompts users that the flag exists though.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by pablo-lua** [comment]
> Hmm, if we add a dev-dependency on `bevy` in the `bevy` `Cargo.toml` with `bevy_debug_tools`  enabled does that work?
> 
> This may still be better, as it prompts users that the flag exists though.

That does work, but that can be strange. If for some reason you enable the feature, the dev_tools plugin will be auto Added by the feature in `DefaultPlugins`, so I thought it would be strange to have the "dependencies" duplicated in examples
If that's okay, I can change it though, but like you said, there is value in have it as a feature IMO

**Comment by alice-i-cecile** [comment]
Yeah, I think this is better. Thanks for checking!

**Comment by pablo-lua** [comment]
> You added a new example but didn't add metadata for it. Please update the root Cargo.toml file.

Third time in this PR.

**Comment by pablo-lua** [comment]
@alice-i-cecile Should be alright now, main Ci passed.

**Pull request merged**
